File: sample-objc/sample-objc/Controller/STInterstitialAdViewController.m
//
//  STInterstitialAdViewController.m
//  santa-ios-sdk-sample-obj
//
//  Created by Jaeuk Jeong on 1/30/24.
//

#import "STInterstitialAdViewController.h"
#import <SantaSDK/SantaSDK-Swift.h>

@interface STInterstitialAdViewController ()<STInterstitialAdViewDelegate>

@property (weak, nonatomic) IBOutlet UITextField *textKeyword;
@property (weak, nonatomic) IBOutlet UIButton *btnLoadAd;
@property (weak, nonatomic) IBOutlet UIButton *btnShowAd;

@property (nonatomic, strong) STInterstitialAdView *interstitial;
@property (nonatomic, strong) NSEnumerator<NSString *> *mediations;

@end

@implementation STInterstitialAdViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    self.textKeyword.text = self.info.ID;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)didTapLoadButton:(id)sender
{
    [self.textKeyword endEditing:YES];

    self.btnShowAd.hidden = NO;

    self.interstitial = [STInterstitialAdView interstitialAdControllerForAdUnitId:self.textKeyword.text];

    self.interstitial.delegate = self;
    self.interstitial.testing = YES;
    [self.interstitial loadAd];
}

- (IBAction)didTapShowButton:(id)sender
{
    [self.interstitial showFromViewController:self];
}

- (void)loadMediation
{
    // 미디에이션 목록이 존재할때
    if (self.mediations != nil) {
        // 미디에이션 목록을 순차로 가져온다
        NSString *mediation = [self.mediations nextObject];
        
        NSLog(@"loadMediation = %@", mediation);
        
        // 미디에이션이 비어있다면 목록을 모두 순회함
        if (mediation != nil) {
            if ([mediation isEqual: @"Mediation Name"]) {
                // 사용할 미이데이션에 맞는 광고 호출 코드를 작성
            }
        } else {
            NSLog(@"Mediation is Empty");
        }
    } else {
        NSLog(@"Mediation is Empty");
    }
}

#pragma mark - <EBInterstitialAdControllerDelegate>

- (void)interstitialDidLoadAd:(STInterstitialAdView *)interstitial
{
}

- (void)interstitialDidFailToLoadAd:(STInterstitialAdView *)interstitial
{
}

- (void)interstitialDidExpire:(STInterstitialAdView *)interstitial
{
    self.btnShowAd.hidden = YES;
}

- (void)interstitialWillAppear:(STInterstitialAdView *)interstitial
{
}

- (void)interstitialDidAppear:(STInterstitialAdView *)interstitial
{
}

- (void)interstitialWillDisappear:(STInterstitialAdView *)interstitial
{
}

- (void)interstitialDidDisappear:(STInterstitialAdView *)interstitial
{
    self.btnShowAd.hidden = YES;
}

- (void)interstitialDidReceiveTapEvent:(STInterstitialAdView *)interstitial
{
}

- (void)interstitialDidLoadMediation:(NSArray<NSString *> *)mediations
{
    self.mediations = [mediations objectEnumerator];
    [self loadMediation];
}

@end

File: sample-swift/sample-swift/Controller/STInterstitialAdViewController.swift
//
//  STInterstitialAdViewController.swift
//  sample-swift
//
//  Created by Jaeuk Jeong on 12/15/23.
//

import UIKit
import SantaSDK

class STInterstitialAdViewController: UIViewController {

    @IBOutlet var textKeyword: UITextField!
    @IBOutlet var btnLoadAd: UIButton!
    @IBOutlet var btnShowAd: UIButton!

    var info: STAdInfoModel?
    var interstitial: STInterstitialAdView?

    override func viewDidLoad() {
        super.viewDidLoad()

        textKeyword.text = info?.ID
    }
}

extension STInterstitialAdViewController {
    @IBAction func loadAdClick(_ sender: UIButton) {

        self.textKeyword.resignFirstResponder()

        self.btnShowAd.isEnabled = true

        if let adUnitId = textKeyword.text {
            self.interstitial = STInterstitialAdView.interstitialAdControllerForAdUnitId(adUnitId)

            if let interstitial = interstitial {
                interstitial.delegate = self        // 대리자 전달
                interstitial.adFormat = [.HTML]     // 광고 포맷 선택 (adFormat: [SantaSDK.STAdFormat]?)
                interstitial.testing = true         // 광고 테스트 여부
                interstitial.loadAd()
            }
        }
    }

    @IBAction func showAdClick(_ sender: UIButton) {
        self.interstitial?.showFromViewController(self)
    }
}

extension STInterstitialAdViewController: STInterstitialAdViewDelegate {
    /**
     * 전면 광고를 성공적으로로드하면 전송됩니다.
     */
    func interstitialDidLoadAd(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidLoadAd")
    }

    /**
     * 광고를로드하지 못할 때 전송됩니다..
     */
    func interstitialDidFailToLoadAd(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidFailToLoadAd")

    }

    /**
     * 전면 광고가 화면에 표시되기 직전에 전송됩니다.
     */
    func interstitialWillAppear(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialWillAppear")
    }

    /**
     * 전면 광고가 화면에 표시된 후에 전송됩니다.
     */
    func interstitialDidAppear(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidAppear")
    }

    /**
     * 전면 광고가  화면에 표시되지 못할때
     */
    func interstitialDidFailToShow(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidFailToShow")
    }

    /**
     * 전면 광고가  화면에서 닫히기 직전에 전송됩니다.
     */
    func interstitialWillDisappear(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialWillDisappear")
    }

    /**
     * 전면 광고가 화면에서 해제 된 후 전송되어 애플리케이션에 제어권이 반환됩니다.
     */
    func interstitialDidDisappear(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidDisappear")
    }

    /**
     * 로드 된 전면 광고를 더 이상 표시 할 수 없을 때 전송됩니다.
     */
    func interstitialDidExpire(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidExpire")
    }

    /**
     * 사용자가 전면 광고를 탭하고 광고가 타겟 작업을 수행하려고 할 때 전송됩니다.
     */
    func interstitialDidReceiveTapEvent(_ interstitial: SantaSDK.STInterstitialAdView?) {
        print("interstitialDidReceiveTapEvent")
    }
}
